formats/src/main/scala/overflowdb/formats/neo4jcsv/Neo4jCsvImporter.scala

```diff
@@ -0,0 +1,244 @@
+      val propertyDef = entry match {
+        case ColumnType.LabelMarker  =>
+          if (labelColumnFound)
+            throw new NotImplementedError(s"multiple ${ColumnType.Label} columns found in $headerFile, which is not supported by overflowdb")
+          labelColumnFound = true
+          fileType = Option(FileType.Nodes)
+          ColumnDef(None, ColumnType.Label)
+        case ColumnType.TypeMarker =>
+          fileType = Option(FileType.Relationships)
+          ColumnDef(None, ColumnType.Type)
+        case s if s.endsWith(ColumnType.Id.toString) =>
+          ColumnDef(None, ColumnType.Id)
+        case s if s.endsWith(ColumnType.StartId.toString) =>
+          ColumnDef(None, ColumnType.StartId)
+        case s if s.endsWith(ColumnType.EndId.toString) =>
+          ColumnDef(None, ColumnType.EndId)
+        case propertyDef if propertyDef.contains(":") =>
+          val name :: valueTpe0 :: Nil = propertyDef.split(':').toList
+          val isArray = propertyDef.endsWith(ColumnType.ArrayMarker) // from the docs: "To define an array type, append [] to the type"
+          val valueTpe =
+            if (isArray) valueTpe0.dropRight(2)
+            else valueTpe0
+          ColumnDef(Option(name), valueType = ColumnType.withName(valueTpe), isArray)
+        case propertyName =>
+          ColumnDef(Option(propertyName), valueType = ColumnType.String) // if property is not annotated with `:someType`, default to String
+      }
+      propertyDefs.addOne((idx, propertyDef))
+    }
+
+    val propertyDefsRes = propertyDefs.result()
+    fileType match {
+      case Some(FileType.Nodes) =>
+        assert(propertyDefsRes.values.exists(_.valueType == ColumnType.Id),
+          s"no :ID column found in headerFile $headerFile - see format definition in $Neo4jAdminDoc")
+        ParsedHeaderFile(FileType.Nodes, propertyDefsRes)
+      case Some(FileType.Relationships) =>
+        assert(propertyDefsRes.values.exists(_.valueType == ColumnType.StartId),
+          s"no :START_ID column found in headerFile $headerFile - see format definition in $Neo4jAdminDoc")
+        assert(propertyDefsRes.values.exists(_.valueType == ColumnType.EndId),
+          s"no :END_ID column found in headerFile $headerFile - see format definition in $Neo4jAdminDoc")
+        ParsedHeaderFile(FileType.Relationships, propertyDefsRes)
+      case _ =>
+        throw new AssertionError(s"unable to determine file type - neither ${ColumnType.Label} (for nodes) nor ${ColumnType.Type} (for relationships) found")
+    }
+  }
+
+  private def parseNodeRowData(columnsRaw: Seq[String], lineNo: Int, columnDefs: Map[Int, ColumnDef]): ParsedNodeRowData = {
+    var id: Integer = null
+    var label: String = null
+    val properties = Seq.newBuilder[ParsedProperty]
+    columnsRaw.zipWithIndex.foreach { case (entry, idx) =>
+      assert(columnDefs.contains(idx), s"column with index=$idx not found in column definitions derived from headerFile")
```

Comment:
I would consider:

```
      columnDefs.get(idx) match {
        case Some(ColumnDef(_, ColumnType.Id, _)) =>
          id = entry.toInt
        case Some(ColumnDef(_, ColumnType.Label, _)) =>
          label = entry
        case Some(ColumnDef(Some(name), valueType, false)) =>
          parseProperty(entry, name, valueType).foreach(properties.addOne)
        case Some(ColumnDef(Some(name), valueType, true)) =>
          parseArrayProperty(entry, name, valueType).foreach(properties.addOne)
        case Some(other) =>
          throw new MatchError(s"unhandled case $other in $lineNo: ${columnsRaw.mkString(",")}")
        case None =>
          throw new AssertionError(s"column with index=$idx not found in column definitions derived from headerFile")
      }
```

It's bit more idiomatic for my taste. Also, it uses just one Map lookup instead of two (`Map.contains` and `Map.apply` both need to look up for the key)

Reply:
Hmm, I started with what you're proposing here, but then changed it around to the current version, which I found more readable. There's already a lot of clutter in decomposing `ColumnDef`, if you add `Option` to the mix it starts getting a bit hard to digest. Matter of taste I guess. 

tl;dr; I see your point but will give this one a miss :)